### ISSUE TITLE
Ungraceful termination when running kube-proxy in IPVS mode

### ISSUE THREAD

**Issue by aaronbbrown** [kubernetes/kubernetes#99162]
#### What happened: 

Connections time out from external client when pods are deleted and kube-proxy is in IPVS mode. 

#### What you expected to happen:

Connections should not timeout or be blackholed when pods are deleted in an IPVS-enabled cluster.

#### How to reproduce it (as minimally and precisely as possible):

This configuration should facilitate graceful termination behavior when pods are deleted w/ `kubectl delete`

<details><summary>nginx K8s YAML</summary>

```yaml
---
apiVersion: v1
kind: Service
metadata:
  name: nginx
  labels:
    app: nginx
spec:
  ports:
  - port: 80
    name: nginx
    targetPort: http
  type: NodePort
  selector:
    app: nginx
---
apiVersion: apps/v1
kind: Deployment
metadata:
  name: nginx
spec:
  strategy:
    rollingUpdate:
      maxSurge: 100%
  selector:
    matchLabels:
      app: nginx
  replicas: 10
  template:
    metadata:
      labels:
        app: nginx
    spec:
      containers:
      - name: nginx
        image: nginx:latest
        ports:
        - containerPort: 80
          name: http
        readinessProbe:
          httpGet:
            path: /
            port: 80
            scheme: HTTP
        lifecycle:
          preStop:
            exec:
              command:
              - /bin/sh
              - -c
              - "sleep 5; /usr/sbin/nginx -s quit"
```

</details>

I verified that requests to an iptables-enabled cluster continued without error while periodically deleting 1 of 10 healthy Pods (using `cyclepods.sh` 👇🏻 ) while a `vegeta attack` was running against the cluster (using `bench.sh` 👇🏻).

<details><summary>bench.sh</summary>

```shell
#!/usr/bin/env bash

set -e

export PATH=$PATH:$HOME/bin
CLUSTER=$1
if [ -z "$CLUSTER" ]; then
  echo "usage: $0 CLUSTER"
  exit 1
fi


kubectl() {
  command kubectl --context "$CLUSTER" -o json "$@"
}

node() {
  kubectl -l kubernetes.github.com/role=node get nodes |
    jq -er '.items[0].metadata.name'
}

serviceport() {
  kubectl -n default get services nginx |
    jq -er '.spec.ports[] | select(.targetPort=="http").nodePort'
}

bench() {
  echo "$1" |
    prlimit -n65535 \
      vegeta attack -duration 5m -keepalive=false -rate '100/1s' -workers 10 |
      vegeta report
}

main() {
  local node=$(node)
  local ip=$(dig +short "$node" | head -1)
  local serviceport=$(serviceport)
  local request="GET http://${ip}:${serviceport}/"

  bench "$request"
}

main
```
</details>

<details><summary>cyclepods.sh</summary>

```shell

#!/usr/bin/env bash

set -e

CLUSTER=$1
if [ -z "$CLUSTER" ]; then
  echo "usage: $0 CLUSTER"
  exit 1
fi

while : ; do
  kubectl --context "$CLUSTER" wait --for=condition=available deployment/nginx
  kubectl --context "$CLUSTER" get pods -l app=nginx -n default --no-headers |
    awk -F' *|/' '$2 == $3 && $4 == "Running" {print $1}' |
    shuf |
    head -1 |
    xargs kubectl --context "$CLUSTER" delete pod
  sleep 0.5
done
```

</details>

Running these script pairs against the iptables-enabled cluster yielded expected results: 100% success.

```console
$ ./bench.sh iptables-cluster
Requests      [total, rate, throughput]         30000, 100.00, 100.00
Duration      [total, attack, wait]             5m0s, 5m0s, 2.622ms
Latencies     [min, mean, 50, 90, 95, 99, max]  1.221ms, 2.433ms, 1.782ms, 3.223ms, 3.964ms, 12.465ms, 85.921ms
Bytes In      [total, mean]                     18360000, 612.00
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           100.00%
Status Codes  [code:count]                      200:30000
Error Set:
```

However, running this against the IPVS-enabled cluster, yielded a lot of timeouts:

```console
$ ./bench.sh ipvs-cluster
Requests      [total, rate, throughput]         30000, 100.00, 79.93
Duration      [total, attack, wait]             5m30s, 5m0s, 29.82s
Latencies     [min, mean, 50, 90, 95, 99, max]  2.597ms, 3.623s, 5.365ms, 30s, 30s, 30s, 30.001s
Bytes In      [total, mean]                     16132932, 537.76
Bytes Out     [total, mean]                     0, 0.00
Success       [ratio]                           87.87%
Status Codes  [code:count]                      0:3639  200:26361
Error Set:
Get "http://10.0.32.39:31641/": dial tcp 0.0.0.0:0->10.0.32.39:31641: connect: connection refused
Get "http://10.0.32.39:31641/": context deadline exceeded (Client.Timeout exceeded while awaiting headers)
Get "http://10.0.32.39:31641/": read tcp 10.52.122.12:39881->10.0.32.39:31641: read: connection reset by peer
```

#### Anything else we need to know?:

We have the following sysctl's set for Linux Virtual Server:
```
net.ipv4.vs.am_droprate = 10
net.ipv4.vs.amemthresh = 1024
net.ipv4.vs.backup_only = 0
net.ipv4.vs.cache_bypass = 0
net.ipv4.vs.conn_reuse_mode = 0
net.ipv4.vs.conntrack = 1
net.ipv4.vs.drop_entry = 0
net.ipv4.vs.drop_packet = 0
net.ipv4.vs.expire_nodest_conn = 1
net.ipv4.vs.expire_quiescent_template = 1
net.ipv4.vs.ignore_tunneled = 0
net.ipv4.vs.nat_icmp_send = 0
net.ipv4.vs.pmtu_disc = 1
net.ipv4.vs.schedule_icmp = 0
net.ipv4.vs.secure_tcp = 0
net.ipv4.vs.sloppy_sctp = 0
net.ipv4.vs.sloppy_tcp = 0
net.ipv4.vs.snat_reroute = 1
net.ipv4.vs.sync_persist_mode = 0
net.ipv4.vs.sync_ports = 1
net.ipv4.vs.sync_qlen_max = 1028346
net.ipv4.vs.sync_refresh_period = 0
net.ipv4.vs.sync_retries = 0
net.ipv4.vs.sync_sock_size = 0
net.ipv4.vs.sync_threshold = 3	50
net.ipv4.vs.sync_version = 1
```

I have reviewed the following issues and it seems like our sysctls are reasonable:
- https://github.com/kubernetes/kubernetes/issues/81775
- https://github.com/kubernetes/kubernetes/issues/71358

`kube-proxy` is configured with the following parameters:

```
kube-proxy --kubeconfig=/etc/kubernetes/kubelet.conf --cluster-cidr=10.125.0.0/17 --v=4 --proxy-mode=ipvs --ipvs-scheduler=rr --conntrack-max-per-core=0
```

#### Environment:
- Kubernetes version (use `kubectl version`): 1.20.2
- Cloud provider or hardware configuration: Azure
- OS (e.g: `cat /etc/os-release`): Debian Stretch
- Kernel (e.g. `uname -a`): 4.19.160-2~deb9u1
- Install tools:
- Network plugin and version (if this is a network-related bug): canal (Calico v3.16.4, Flannel v0.12.0)
- Others:

**Comment by uablrek** [comment]
@aaronbbrown Excellent written issue :+1: Thanks!

I can reproduce this as described.

/triage accepted

I would like to investigate this a bit further so I don't unassign myself just yet.

**Comment by aaronbbrown** [comment]
Thanks, @uablrek! ❤️  Please let me know if you need any more details about our configuration.

**Comment by uablrek** [comment]
To see the dynamics do `watch ipvsadm -Ln -t 192.168.1.2:32502` (with your address of course). Below is a snapshot but to watch it unfold in RT is more interresting;
```
$ ipvsadm -Ln -t 192.168.1.2:32502
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  192.168.1.2:32502 rr
  -> 11.0.1.8:80                  Masq    0      3          555       
  -> 11.0.2.7:80                  Masq    1      0          599       
  -> 11.0.2.9:80                  Masq    1      0          599       
  -> 11.0.2.10:80                 Masq    0      48         99        
  -> 11.0.2.11:80                 Masq    1      0          252       
  -> 11.0.3.7:80                  Masq    0      63         269       
  -> 11.0.3.8:80                  Masq    1      0          599       
  -> 11.0.3.9:80                  Masq    0      58         186       
  -> 11.0.3.10:80                 Masq    1      0          523       
  -> 11.0.3.11:80                 Masq    1      0          199       
  -> 11.0.3.12:80                 Masq    1      0          106       
  -> 11.0.4.7:80                  Masq    1      0          599       
  -> 11.0.4.8:80                  Masq    0      11         25        
  -> 11.0.4.9:80                  Masq    0      61         357       
  -> 11.0.4.10:80                 Masq    0      22         230       
  -> 11.0.4.11:80                 Masq    1      0          43        
```
From a first glance it looks like the ipvs-proxier does it right: it sets Weight=0 for the deleted pods but,
_the `ActiveConn` is incremented even for pods with Weight=0._

This seems wrong to me. I will hopefully get some time to investigate further today.

**Comment by uablrek** [comment]
I changed the scheduler to "wrr", weighted round-robin, but the result is the same. Kind of a relief since to my experience Weight=0 is honored for scheduler "rr".

**Comment by uablrek** [comment]
I was't able to check this during the week, but here are some new observations.

My first thought was that this is some kind of race; when a pod is deleted some SYNs may "sneak through" before weight=0 is set leaving states in conntrack and ipvs and what-not.

### The pods don't have to be deleted with traffic running

1. Run a `vegeta attack` for a short time to get some "InActConn"s in ipvs, then stop it (^C)
2. Delete a POD. The entry is still in ipvs but weight=0
3. Start a new `vegeta attack`

Soon the entry with weight=0 will get some "ActiveConn"s and some traffic is lost.

Commands;
```
nodeport=$(kubectl get svc nginx -o json | jq -r .spec.ports[].nodePort)
echo "GET http://192.168.1.2:$nodeport" \
 | vegeta attack -duration 1m -keepalive=false -rate '100/1s' -workers 10 \
 | vegeta report
pod=$(kubectl get pods -l app=nginx -o name | shuf | head -1)
kubectl delete $pod
```

### There are many sockets in SYN_SENT

On the sending node;
```
$ netstat -putn
Proto Recv-Q Send-Q Local Address           Foreign Address         State       PID/Program name    
tcp        0      1 192.168.1.201:58749     192.168.1.2:31634       SYN_SENT    491/vegeta          
tcp        0      1 192.168.1.201:39159     192.168.1.2:31634       SYN_SENT    491/vegeta          
..
```
Note that the `Send-Q` is 1.

On a test with `bench.sh` and `cyclepods.sh` the number of open sockets in SYN_SENT becomes very large, hence the `prlimit` in `bench.sh`. I don't have that  program on my test system but `ulimit` works just as well. Without it you will get "too many open files".

### Problem observed with another traffic generator

I tried to generate traffic with [kahttp](https://github.com/Nordix/kahttp) which I know better since it's my own program. I had to speed up the rate but then the same problem was observed; `ActiveConn` was incremented for entries with weight=0.

```
nodeport=$(kubectl get svc nginx -o json | jq -r .spec.ports[].nodePort)
kahttp -disable_ka -monitor -nclients 20 -rate 400 -timeout 40s -address http://192.168.1.2:$nodeport | jq .
...
{
  "Started": "2021-02-27T08:19:28.32452116Z",
  "Duration": 60175230340,
  "Rate": 400,
  "Clients": 20,
  "Dials": 5973,
  "FailedConnections": 20,
  "Sent": 5953,
  "Received": 5953,
  "Dropped": 2710,
  "FailedConnects": 0,
  "NotOK": 0
}
```
`FailedConnections: 20` corresponds exactly with the oberved sum of `ActiveConn` for entries with weight=0.

"FailedConnections" means that the TCP connect was succesful but the request was not. "FailedConnects" counts the failed TCP connects.

**Comment by uablrek** [comment]
I begin to suspect a kernel problem. I upgraded my kernel to `linux-5.11.1` but the problem is still there.

Another observation; when weight=0 is set for an entry it receives no new connections, `ActiveConn` is zero;
```
# ipvsadm -Ln -t 192.168.1.2:30232
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  192.168.1.2:30232 rr
  -> 11.0.1.9:80                  Masq    1      1          671       
  -> 11.0.1.10:80                 Masq    1      1          671       
  -> 11.0.2.8:80                  Masq    1      0          672       
  -> 11.0.2.9:80                  Masq    1      0          672       
  -> 11.0.3.7:80                  Masq    1      0          672       
  -> 11.0.3.8:80                  Masq    1      0          672       
  -> 11.0.3.9:80                  Masq    1      0          672       
  -> 11.0.4.8:80                  Masq    1      0          672       
  -> 11.0.4.9:80                  Masq    0      0          329       
  -> 11.0.4.10:80                 Masq    1      0          672       
```

No problems until...

when the `InActConn` for the entry begin to timeout is seems that it "leaves room" for `ActiveConn`;
```
# ipvsadm -Ln -t 192.168.1.2:30232
Prot LocalAddress:Port Scheduler Flags
  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
TCP  192.168.1.2:30232 rr
  -> 11.0.1.9:80                  Masq    1      0          1425      
  -> 11.0.1.10:80                 Masq    1      0          1425      
  -> 11.0.2.8:80                  Masq    1      0          1425      
  -> 11.0.2.9:80                  Masq    1      0          1425      
  -> 11.0.2.10:80                 Masq    1      0          509       
  -> 11.0.3.7:80                  Masq    1      0          1425      
  -> 11.0.3.8:80                  Masq    1      0          1425      
  -> 11.0.3.9:80                  Masq    1      0          1425      
  -> 11.0.4.8:80                  Masq    1      0          1425      
  -> 11.0.4.9:80                  Masq    0      16         313       
  -> 11.0.4.10:80                 Masq    1      0          1425      
```
Now 16 of the original 329 `InActConn` has timed out and are immediately taken as `ActiveConn` (313+16=329).

**Comment by uablrek** [comment]
I sent an issue on linux;
https://bugzilla.kernel.org/show_bug.cgi?id=211979

**Comment by zq-david-wang** [comment]
Any chance the package got send on default route? Try
`tcpdump -i [default route interface] host [service ip]`
Any package referring to service ip shows up on default interface would indicating something about conntrack (or something else in kernel) wrong.

**Comment by uablrek** [comment]
/cc @andrewsykim 

@aojea pointed out that you have updated ipvs in the kernel in; https://github.com/torvalds/linux/commit/35dfb013149f74c2be1ff9c78f14e6a3cd1539d1

Do you agree that this may be a kernel bug?

**Comment by uablrek** [comment]
This problem is probably related to #81775.

**Comment by uablrek** [comment]
This problem is fixed with a kernel patch by @andrewsykim and included in linux >=5.9, please see https://github.com/kubernetes/kubernetes/pull/102122.

I have made a PR https://github.com/kubernetes/kubernetes/pull/102122 that leaves net.ipv4.vs.conn_reuse_mode=1 for kernels >=5.9.

I have successfully tested with the PR applied on linux-5.12.1.

**Comment by uablrek** [comment]
When https://github.com/kubernetes/kubernetes/pull/102122 is merged I will close this issue. I marked my kernel bug-report as resolved/patch-already-exists.

@aaronbbrown Sorry for the not very user-friendly fix since you have to upgrade the kernel. But working with kernel functions, that's the way it is.

**Comment by aaronbbrown** [comment]
Thanks, @uablrek !

**Comment by uablrek** [comment]
https://github.com/kubernetes/kubernetes/pull/102122 is now merged. This will fix the problem if a kernel >= 5.9 is used.

/close
